hub/integrations/pytorch/dataset.py

```diff
@@ -501,29 +510,28 @@ def __iter__(self):
             while True:
                 next_batch = next(it)
                 if isinstance(next_batch, dict):
-                    d = {}
-                    for k, v in next_batch.items():
-                        current_val = v[0]
-                        if isinstance(current_val, torch.Tensor):
-                            current_val = current_val.clone().detach()
-                        d[k] = current_val
-                    val = IterableOrderedDict(d)
+                    batch_keys = list(next_batch.keys())
+                    vals = (
+                        IterableOrderedDict(
+                            {k: process(next_batch[k][i]) for k in batch_keys}
+                        )
+                        for i in range(len(next_batch[batch_keys[0]]))
+                    )
                 elif isinstance(next_batch, Sequence):
-                    val = []
-                    for item in next_batch:
-                        if isinstance(item, torch.Tensor):
-                            item = item.clone().detach()
-                        val.append(item)
-                else:
-                    val = next_batch
-
-                if buffer is not None:
-                    result = buffer.exchange(val)
-
-                    if result:
-                        yield result
+                    num_samples = len(next_batch[0])
+                    vals = (
+                        [process(next_batch[i][j]) for i in range(len(next_batch))]
+                        for j in range(num_samples)
+                    )
                 else:
-                    yield val
+                    raise ValueError(f"Unsupported sample type: {type(next_batch)}")
```

Comment:
Can this error message be made more useful by mentioning the expected type? "Expected batch to be of type dict but received.. " Etc.